typing/typecore.ml

```diff
@@ -4649,15 +4651,26 @@ and type_let
     )
     pat_list
     (List.map2 (fun (attrs, _) e -> attrs, e) spatl exp_list);
+  let pvs = List.map (fun pv -> { pv with pv_type = instance pv.pv_type}) pvs in
   end_def();
   List.iter2
     (fun pat exp ->
        if not (is_nonexpansive exp) then
-         iter_pattern (fun pat -> generalize_expansive env pat.pat_type) pat)
+         generalize_expansive env pat.pat_type)
     pat_list exp_list;
-  List.iter
-    (fun pat -> iter_pattern (fun pat -> generalize pat.pat_type) pat)
-    pat_list;
+  iter_pattern_variables_type generalize pvs;
+  (* The next line changes the toplevel experience from:
+     {[
```

Comment:
This comment is a bit verbose for a function that is already (too) long.
Couldn't you reduce this to 2 lines (maybe with a pointer to the issue or PR)?
```
(* The next line is needed to generalize toplevel expressions that are not bound to a variable *)
```

Reply:
Personally I would rather have a nice, clear and self-contained comment, even if it takes 10 more lines. (The function, before the change, appears to be 242 lines long, so there is no noticeable change in readability.) If we want a shorter function, maybe we could instead split some parts into auxiliary functions?

Reply:
This may be a question of taste, but if we start to add a 10 line comment each time we fix an issue, the code will soon be unreadable.
In this case, I'm actually the one who asked for the comment, so I think it is useful,  but I believe it could be shorter without losing information.
(The question of how we could split or streamline `type_let` starts to make sense too; it used to be 24 lines in ocaml 2.04.)

Reply:
Here would be my favorite version for this comment:

```ocaml
(* We also generalize expressions that are not bound to a variable.
  This does not matter in general, but those types are shown
  by the interactive toplevel, for example: {[
    let _ = Array.get;;
     - : 'a array -> int -> 'a = <fun>
   ]} *)
```

Reply:
I'm fine with @gasche latest suggestion. I'll make the change.

Reply:
I agree with @gasche here. Adding indirections in the comments is extremely painful to navigate when reading, and it doesn't prevent comment bitrot at all.